node/src/components/linear_chain_sync.rs

```diff
@@ -445,12 +450,16 @@ where
                     self.block_downloaded(rng, effect_builder, block.header())
                 }
                 BlockByHeightResult::FromPeer(block, peer) => {
-                    if block.height() != block_height {
+                    if block.height() != block_height
+                        || *block.header().parent_hash() != self.current_block().unwrap().hash()
```

Comment:
Looks like `block_height` should equal `self.current_block().unwrap().height()`. Instead of doing those to checks, can you replace it with `block != self.current_block.unwrap()`?